Consensus module

DavidVorick opened the pull request:
There's some organizational stuff left but I'l leave it for another PR.

I gave the currency type a decent amount of attention, and I have high confidence in the current code for it.

lukechampine commented on types/currency.go:
```diff
@@ -1,24 +1,44 @@
+// directly to other objects and packages. The currency object should never
+// have a negative value. The currency should never overflow. There is a
+// maximum size value that can be encoded (around 10^600), however exceeding
+// this value will not result in overflow.
```
with the marshalling changes, the maximum value is now approx. 10^10^20

lukechampine commented on types/currency.go:
```diff
@@ -28,34 +48,27 @@ func NewCurrency64(x uint64) (c Currency) {
-	y.i.Sub(&c.i, &x.i)
+// Div returns a new Currency value y = c / x.
+func (c Currency) Div(x Currency) (y Currency) {
+	y.i.Div(&c.i, &x.i)
```
alphabetizing? don't really see the gain here but w/e.

DavidVorick replied:
It's how the go standard library does it. It makes finding functions easier when looking at a file.

lukechampine replied:
Sometimes they do, sometimes they don't. The math.Big arithmetic is grouped together, and the formatting functions are grouped together as well: http://golang.org/src/math/big/int.go?s=2391:2424#L100

I can see the advantage for alphabetizing large files, but in this case it's very obvious that Add, Sub, Mul, and Div are all the same "type" of function and belong together. Besides, you didn't separate Marshal and Unmarshal :P

DavidVorick replied:
life is hard.

I left Marshal and Unmarshal together because it seemed really natural. New is at the top.

lukechampine commented on types/currency.go:
```diff
@@ -1,24 +1,44 @@
+		if DEBUG {
+			panic("cannot have a negative currency")
+		}
+	}
```
you can use the same style you used for MulFloat:
```go
if b.Sign() < 0 {
    if DEBUG {
        panic("cannot have a negative currency")
    }
} else {
    c.i = *b
}
return
```

lukechampine commented on types/currency.go:
```diff
@@ -89,54 +111,66 @@ func (c Currency) Sqrt() (y Currency) {
+	if err != nil {
+		return err
+	}
+	if c.Cmp(ZeroCurrency) < 0 {
```
I would use `c.i.Sign() < 0` for clarity

DavidVorick replied:
There is no Sign() function anymore since it can't ever be negative.

I guess I should add it back though. Cmp(ZeroCurrency) is a pretty awful construction, especially when you have to import the types package just to do that.

lukechampine replied:
Not `Currency.Sign()`, `Int.Sign()`

DavidVorick replied:
ohhhhhhh. Got it.

lukechampine commented on types/currency.go:
```diff
@@ -89,54 +111,66 @@ func (c Currency) Sqrt() (y Currency) {
+	if err != nil {
+		return err
+	}
+	if c.Cmp(ZeroCurrency) < 0 {
```
ditto here

lukechampine commented on types/currency_test.go:
```diff
@@ -0,0 +1,205 @@
+	c := NewCurrency64(7135)
+	if b.String() != c.String() {
+		t.Error("string function not behaving as expected")
+	}
```
seems kinda silly to test the pass-through functions, but I guess the coverage % is nice.
here you can just use `NewCurrency64(7135).String() != "7135"`

DavidVorick replied:
I actually caught an error in one of the pass-through functions, so it's not completely silly.

lukechampine commented on types/currency_test.go:
```diff
@@ -0,0 +1,205 @@
+		if r == nil {
+			t.Error("no panic occured when trying to create a negative currency")
+		}
+	}()
```
you can abstract away the recovery by using a closure:
```go
func catchErr(fn func()) (err error) {
    defer func() {
        if r := recover(); r != nil {
            err = fmt.Errorf("%v", r)
        }
    }()
    fn()
}

func TestNegativeCurrencySub(t *testing.T) {
    err := catchErr(func() {
        NewCurrency64(1).Sub(NewCurrency64(2))
    })
    if err == nil {
        t.Fatal("no panic occurred when using improper arguments to Sub")
    }
}
```

lukechampine commented on types/currency.go:
```diff
@@ -65,23 +78,32 @@ func (c Currency) Mul(x Currency) (y Currency) {
+func (c Currency) MulFloat(x float64) (y Currency) {
+	if x < 0 {
+		if DEBUG {
+			panic("cannot multiple currency by a negative number")
```
multiply

DavidVorick commented:
Sorry for being disorganized, but I fixed all the errors in the 'types-cleanup' branch. The reason is that since I split consensus into two packages, all the imports and stuff had changed and I wanted to avoid merge conflicts as much as possible.

lukechampine commented:
You shouldn't be scared of a simple rebase like this. It's worth the effort to keep things organized.

DavidVorick commented:
\>.< just tried but somehow lost the commit entirely.

=/

lukechampine commented:
lol. I cherry-picked it into this branch locally, I can push it if you want

DavidVorick commented:
oh thanks. That would be helpful
